mira/sources/vensim.py

```diff
@@ -0,0 +1,360 @@
+https://www.vensim.com/documentation/sample_models.html
+
+SIR mdl file example here: https://www.vensim.com/documentation/sample_models.html
+"""
+
+__all__ = ["template_model_from_mdl_file_url"]
+
+import pandas as pd
+import pysd
+from pysd.translators.vensim.vensim_file import VensimFile
+import requests
+import sympy
+import re
+import tempfile
+
+from mira.metamodel import *
+from mira.metamodel.utils import safe_parse_expr
+from mira.metamodel import Concept, TemplateModel
+from mira.sources.util import (
+    parameter_to_mira,
+    transition_to_templates,
+    get_sympy,
+)
+
+
+NEW_CONTROL_DELIMETER = (
+    " ******************************************************** .Control "
+    "********************************************************"
+)
+CONTROL_VARIABLE_NAMES = {"FINALTIME", "INITIALTIME", "SAVEPER", "TIMESTEP"}
+
+
+def state_to_concept(state) -> Concept:
+    """
+    Create a MIRA Concept from a state
+
+    Parameters
+    ----------
+    state : pd.Series
+        The series that contains state data
+
+    Returns
+    -------
+    :
+        The MIRA concept created from the state
+    """
+    name = state["Py Name"]
+    description = state["Comment"]
+    unit_dict = {
+        "expression": state["Units"].replace(" ", "")
+        if state["Units"]
+        else None
+    }
+    unit_expr = get_sympy(unit_dict, UNIT_SYMBOLS)
+    units_obj = Unit(expression=unit_expr) if unit_expr else None
+
+    return Concept(name=name, units=units_obj, description=description)
+
+
+def process_expression_text(expr_text, symbols):
+    """
+    Create a sympy expression from a string expression using the supplied mapping of symbols
+
+    Parameters
+    ----------
+    expr_text : str
+        The string expression
+
+    symbols : dict[str,sympy.Symbol]
+        A mapping of string symbol to a symbol in sympy
+
+    Returns
+    -------
+    : sympy.Expr
+        The sympy expression
+    """
+    # strip leading and trailing white spaces
+    # remove spaces between operators and operands
+    # replace space between two words that makeup a variable name with "_"'
+    expr_text = (
+        expr_text.strip()
+        .replace(" * ", "*")
+        .replace(" - ", "-")
+        .replace(" / ", "/")
+        .replace(" + ", "+")
+        .replace("^", "**")
+        .replace(" ", "_")
+        .replace('"', "")
+        .lower()
+    )
+
+    sympy_expr = safe_parse_expr(expr_text, symbols)
+    return sympy_expr
+
+
+def template_model_from_mdl_file_url(url) -> TemplateModel:
+    """
+    Return a template model from a Vensim file
+
+    Parameters
+    ----------
+    url : str
+        The url to the mdl file
+
+    Returns
+    -------
+    :
+        A MIRA Template Model
+    """
+    data = requests.get(url).content
+    temp_file = tempfile.NamedTemporaryFile(
+        mode="w+b", suffix=".mdl", delete=False
+    )
+
+    with temp_file as file:
+        file.write(data)
+
+    utf_encoding = "{UTF-8} "
+
```

Comment:
We could refactor this to also add a `template_model_from_mdl_file(fname)` function at this point which is useful for processing local files.